Added Current Sensors, Net Sensors

cabberley opened the pull request:
This pull request has updates to provide more sensors
- adds a sensor for the current for each phase
- adds a Net Import/export sensor for kW and AMPs
- renamed the Usable Battery Capacity sensor to align to the rest of the naming conventions

Tested changes with my ST10000 Hybrid 3 phase

cabberley committed:
Update sensor.py

cabberley committed:
Update hacs.json

cabberley committed:
Merge pull request #1 from juicejuice/main

update

cabberley committed:
Merge branch 'juicejuice:main' into develop

cabberley committed:
Update sensor.py

cabberley committed:
Update redbacklib.py

cabberley committed:
Update hacs.json

cabberley committed:
Merge pull request #2 from cabberley/develop

Added new sensors and renamed entity

cabberley committed:
Update sensor.py

cabberley committed:
Update const.py

cabberley committed:
Update coordinator.py

cabberley committed:
Update redbacklib.py

cabberley committed:
Merge pull request #3 from cabberley/develop

added more sensors

cabberley committed:
Add files via upload

cabberley committed:
Merge pull request #4 from cabberley/develop

Add files via upload

juicejuice commented:
Thanks so much for the contributions! Lots of extra sensors :) I feel like some of the sensors such as serial number should perhaps be device info rather than sensors but that's just a hunch. I'm no HA guru. Also, I wonder if we can have someone test on single phase just to make sure the new sensors work in that scenario.
Two specific things before we merge:
1. The use of "SensorDeviceClass.ENERGY_STORAGE" means we now need minimum HA level of 2023.4 to be set in hacs.json, can you please do that?
2. The battery level is no longer shown correctly in the device info screen. Any chance you can look into a fix?
![image](https://github.com/juicejuice/homeassistant_redback/assets/5318048/36b19de6-e684-408a-8cc1-211d7e71a84d)

cabberley committed:
Update hacs.json

cabberley committed:
Update hacs.json

cabberley committed:
Delete custom_components/redback directory

cabberley committed:
Create strings.json

cabberley committed:
Add files via upload

cabberley committed:
Create en.json

cabberley committed:
Update sensor.py

cabberley committed:
Update sensor.py

cabberley committed:
Merge pull request #5 from cabberley/develop

Develop

cabberley commented:
Thanks for the feedback. I have updated the code and cleaned it up. the HA Version requirement has been updated in the json file as well.

The icon on the device screen, can't find any doco on how that is driven, after working through it, it appears it automatically selects sensors which has percent assigned to it and then displays the one with the lowest value.

So I have shifted a few metrics to extended attributes on Battery Capacity, Inverter Status and Battery SoC. This addresses the problem and also cleans up the sensor list a bit.

You may also like to have a look at this repo that sets up everything to actually control the Hybrid inverter status, it does require you to store your portal email & password, which is being done using secrets file. It does require my main branch to get the right data from the redback integration as there are a couple of items it needs which aren't in your main branch yet.

https://github.com/cabberley/Home-Assistant_Redback_Invertor_Mode_Set

GrantKeymer commented:
Great work Chris Abberley and Juice 👏 

Is there any way we could access Battery SoH?  The website portal under 'System Details' provides a field for overall battery and individual battery SoH (4 fields in my case).  It looks like this...

![Battery 1 Statistics](https://github.com/juicejuice/homeassistant_redback/assets/35156198/d8c96a0c-004e-4308-9274-0dec5d7b5061)

For the past 6 weeks I've been waking up to this...

![Missing Batteries Message](https://github.com/juicejuice/homeassistant_redback/assets/35156198/0a9cabe5-6794-40af-921f-94f97554b9fb)

In early April, I noticed the first subtle changes to my charge and discharge curves which in hindsight, indicated something was not right with my batteries.  By the end of May, Battery 2 was discharging completely to about 2% whilst other batteries remained between 40 and 60%.  When Battery 2 is almost fully discharged, it cuts off the RS485 comms to all remaining batteries in the daisy chain, hence the message that my system has 'Missing Batteries 1/4'.

If I had been able to view the SoH for each battery, I suspect I would have noticed something amiss much earlier.  Battery 2 has now lost most of its capacity and I'm wondering if Battery 3 (the other 2.4kWh module) is headed down the same track.  Looking at the individual discharge curves, I can see that the 2.4kWh batteries are discharging more deeply each night than the other two 3.55kWh batteries.  This is probably why Battery 2 has given out first.

I stumbled across a Pylontech users page on Facebook and they were discussing access to the RS485 comms between batteries.  That is really inconvenient for me, but apparently it allows viewing all the individual battery statistics including SoH for each module.  Is there any way to access SoH via the Public or Private APIs?

cabberley committed:
Update sensor.py

cabberley committed:
Update sensor.py

cabberley committed:
Merge pull request #6 from cabberley/develop

Added additional Battery sensor for charge/discharge flow and current stored capacity

cabberley committed:
Update sensor.py

cabberley committed:
Update sensor.py

cabberley committed:
Update sensor.py

cabberley committed:
Update sensor.py

cabberley committed:
Merge pull request #7 from cabberley/develop

Update sensor.py

cabberley committed:
Update sensor.py

cabberley committed:
Merge pull request #8 from cabberley/develop

Update sensor.py

cabberley committed:
Update sensor.py

fix battery class

cabberley committed:
Squashed commit of the following:

commit 610b9f69f5dd29018075ac4e1ca6c5087c7a72ae
Merge: ad3eacb 203ef21
Author: juicejuice <juicejuice@gmail.com>
Date:   Tue Dec 19 07:04:48 2023 +1000

    Merge pull request #20 from Mark-Hetherington/cumulative-energy

    Measure cumulative energy by maintaining running total

commit 203ef218d8686a4a18fd8ee3d7f6149954dbde47
Author: Mark Hetherington <mark@hethos.org>
Date:   Fri Dec 15 21:47:18 2023 +1100

    Accumulate the measurement after scaling

commit 2d75e0b7ef46aafaab18cd5f93545011568cd3a3
Author: Mark Hetherington <mark@hethos.org>
Date:   Fri Dec 15 11:34:13 2023 +1100

    Add missing parameter for superclass

commit 9849b87f31023a5e7058f19d6a90ff62b6e5997d
Author: Mark Hetherington <mark@hethos.org>
Date:   Fri Dec 15 11:28:38 2023 +1100

    Use a cumulative energy total

commit ad3eacb66b455489c7cf2c327287b270928ebb54
Author: goakes <goakes@gpsl.co>
Date:   Thu Dec 14 16:07:44 2023 +1000

    Fix Site Load sensor when no battery exists. https://github.com/juicejuice/homeassistant_redback/issues/17

commit 7bb7012f4405a21c327bac21b1c8f09f3856b2b9
Author: juicejuice <juicejuice@gmail.com>
Date:   Sun Nov 26 18:25:06 2023 +1000

    Restore missing curly brace

commit 40d3a1c18b8a11347412000c37113f1262b37301
Author: juicejuice <juicejuice@gmail.com>
Date:   Sun Nov 26 16:46:01 2023 +1000

    Support re-authentication of Redback devices when API credentials expire

cabberley committed:
Merge pull request #10 from cabberley/develop

Squashed commit of the following:

cabberley committed:
Update manifest.json

cabberley committed:
Create validate.yaml

cabberley committed:
Delete .github/workflows/hacs.yml

cabberley committed:
Update sensor.py
